lib/ws.c

```diff
@@ -1167,89 +1187,107 @@ CURL_EXTERN CURLcode curl_ws_send(CURL *d, const void *buffer,
   }
 
   /* Not RAW mode, buf we do the frame encoding */
-  space = Curl_bufq_space(&ws->sendbuf);
-  CURL_TRC_WS(data, "curl_ws_send(len=%zu), sendbuf=%zu space_left=%zu",
-              buflen, Curl_bufq_len(&ws->sendbuf), space);
-  if(space < 14) {
-    result = CURLE_AGAIN;
+
+  if(ws->sendbuf_payload && (buflen < ws->sendbuf_payload)) {
+    /* We have been called with LESS buffer data than before. This
+     * is not how it's supposed too work. */
+    failf(data, "curl_ws_send() called with less data then a previous "
+          "call that got CURLE_AGAIN as result.");
+    result = CURLE_BAD_FUNCTION_ARGUMENT;
     goto out;
   }
 
-  if(flags & CURLWS_OFFSET) {
-    if(fragsize) {
-      /* a frame series 'fragsize' bytes big, this is the first */
-      n = ws_enc_write_head(data, &ws->enc, flags, fragsize,
+  if(ws->enc_ack_pending) {
+    /* Flushed a frame that we EAGAINed before */
```

Comment:
I am not sure how this block could work unless a flush call takes place at some point before to continue sending the frame. For example consider a case without raw mode where the first curl_ws_send call is unable to send a frame start so it sets enc_ack_pending TRUE and returns CURLE_AGAIN. Then for the second call I'd expect this block would hit but when it does it signals that the entire payload is sent even though no payload has actually been sent? like this:

curl_ws_send 123 bytes -> (frame send incomplete) -> CURLE_AGAIN
curl_ws_send 123 bytes -> (no flush?) -> CURLE_OK and _sent_ receives 123

Reply:
Thanks for the review. You are correct. There was originally a flush before this, I move it into the loop and now these checks are no longer correct. Working on a fix...

Reply:
Rewrote checks and flushes to assure the correct behaviour. Would be happy to get a review again.